pkg/parquetquery/predicates.go

```diff
@@ -261,6 +261,94 @@ func (p *IntBetweenPredicate) KeepPage(page pq.Page) bool {
 	return true
 }
 
+// Generic predicate with callbacks to evalulate data of type T
+// Fn evalulates a single data point and is required. Optionally,
+// a RangeFn can evalulate a min/max range and is used to
+// skip column chunks and pages when RangeFn is supplied and
+// the column chunk or page also include bounds metadata.
+type GenericPredicate[T any] struct {
+	Fn      func(T) bool
+	RangeFn func(min, max T) bool
+	Extract func(pq.Value) T
+}
+
+var _ Predicate = (*GenericPredicate[int64])(nil)
+
+func NewGenericPredicate[T any](fn func(T) bool, rangeFn func(T, T) bool, extract func(pq.Value) T) *GenericPredicate[T] {
```

Comment:
nit: Should extract return a bool `ok`? That way the calling code can signal that the `pq.Value` cannot be translated to whatever type is being attempted. 

Depending on how this is intended to be used might not matter at all.

Reply:
I think an `ok` is not needed because all values in a column have the same type.  The case where some values convert but others can't seems more like T isn't defined appropriately to the column.   The intent isn't to support any T, ideally only the same data types as parquet (int64, etc), but wasn't sure the best way to implement that constraint in go generics (is it possible?).  Further down this line of thinking, we could eliminate `Extract` altogether and automatically get the right data type based on T.